browser/upgrade_detector/upgrade_when_idle.cc

```diff
@@ -0,0 +1,132 @@
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/first_run/upgrade_util.h"
+#include "chrome/browser/lifetime/application_lifetime.h"
+#include "chrome/browser/profiles/profile_manager.h"
+#include "chrome/browser/ui/browser_list.h"
+#include "chrome/browser/upgrade_detector/upgrade_detector.h"
+#include "chrome/common/chrome_switches.h"
+#include "components/browsing_data/core/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "ui/base/idle/idle.h"
+
+namespace {
+
+// How much idle time (since last input event) must have passed until we
+// restart the browser when an update is available and no window is open.
+constexpr int kIdleAmount = 3;  // Minutes (or seconds, if testing).
+
+bool IsTesting() {
+  // Implementation similar to IsTesting() in upgrade_detector_impl.cc.
+  const base::CommandLine& cmd_line = *base::CommandLine::ForCurrentProcess();
+  return cmd_line.HasSwitch(switches::kSimulateUpgrade) ||
+         cmd_line.HasSwitch(switches::kCheckForUpdateIntervalSec) ||
+         cmd_line.HasSwitch(switches::kSimulateCriticalUpdate) ||
+         cmd_line.HasSwitch(switches::kSimulateOutdated) ||
+         cmd_line.HasSwitch(switches::kSimulateOutdatedNoAU);
+}
+
+bool AreAnyBrowsersOpen() {
+  return BrowserList::GetInstance()->size() > 0;
+}
+
+bool AreAnyClearDataOnExitSettingsEnabled() {
+  ProfileManager* profile_manager = g_browser_process->profile_manager();
+  if (!profile_manager) {
+    // This can happen during tests.
+    return false;
+  }
+  for (Profile* profile : profile_manager->GetLoadedProfiles()) {
+    PrefService* prefs = profile->GetPrefs();
+    // Check if any of the "clear on exit" preferences are enabled
+    if (prefs->GetBoolean(browsing_data::prefs::kDeleteBrowsingHistoryOnExit) ||
+        prefs->GetBoolean(browsing_data::prefs::kDeleteDownloadHistoryOnExit) ||
+        prefs->GetBoolean(browsing_data::prefs::kDeleteCacheOnExit) ||
+        prefs->GetBoolean(browsing_data::prefs::kDeleteCookiesOnExit) ||
+        prefs->GetBoolean(browsing_data::prefs::kDeletePasswordsOnExit) ||
+        prefs->GetBoolean(browsing_data::prefs::kDeleteFormDataOnExit) ||
+        prefs->GetBoolean(browsing_data::prefs::kDeleteHostedAppsDataOnExit) ||
+        prefs->GetBoolean(browsing_data::prefs::kDeleteSiteSettingsOnExit) ||
+        prefs->GetBoolean(browsing_data::prefs::kDeleteBraveLeoHistoryOnExit)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+}  // namespace
+
+namespace brave {
+
+UpgradeWhenIdle::UpgradeWhenIdle() {
+  UpgradeDetector::GetInstance()->AddObserver(this);
+}
+
+UpgradeWhenIdle::~UpgradeWhenIdle() {
+  UpgradeDetector::GetInstance()->RemoveObserver(this);
+}
+
+void UpgradeWhenIdle::OnUpgradeRecommended() {
+  // This function gets called repeatedly when an upgrade is available. When
+  // testing, the interval is every 500ms. In that case, our idle timer needs to
+  // have a shorter interval than that in order to run. We use 250ms.
```

Comment:
use `task_environment_.AdvanceClock` instead

Reply:
"Testing" here doesn't mean running in a browser test; It means running Brave as a user with a switch such as `--check-for-update-interval=10`. In that context, it seems to me that `AdvanceClock` cannot be used; Unless I missed something. Would you please let me know?